torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2623,8 +2627,6 @@ def _cast_forward_inputs(self, *args, **kwargs):
         Moves the :meth:`forward` inputs to the compute device and casts them
         to the appropriate dtype if needed.
         """
-        if not self._is_root:
```

Comment:
This is because otherwise `_cast_forward_inputs ` will not be called in the non recursive policy.

Reply:
IIUC: The issue is that `_lazy_init()` is not called until `_pre_forward()`, which will only be run on the first module (that is not the `FullyShardedDataParallel` module). Then, `_cast_forward_inputs()` returns immediately since `self._is_root` is `None`.

Your change is quite clean 👍🏼